modules/internal/ChapelRange.chpl

```diff
@@ -2176,10 +2176,57 @@ operator :(r: range(?), type t: range(?)) {
   // An unbounded range iterator (for all strides)
   pragma "no doc"
   pragma "order independent yielding loops"
-  iter range.these() where boundedType != BoundedRangeType.bounded {
+  iter range.these() where boundedType == BoundedRangeType.boundedNone {
+    compilerError("iteration over a range with no bounds");
+  }
 
-    if boundedType == BoundedRangeType.boundedNone then
-      compilerError("iteration over a range with no bounds");
+  // An unbounded range iterator (for all strides)
+  pragma "no doc"
+  pragma "order independent yielding loops"
+  iter range.these() where boundedType == BoundedRangeType.boundedLow {
```

Comment:
These two "these" functions have largely identical code and comments.  The main difference I can see is the computation of end and the condition `i > end` vs `i < end` for the "did we wrap around and surpass our representable values" check.  You may want to consider factoring out the common bits of code\comments.

Reply:
I feel disinclined to factor the main "logic" of the loop (from the declaration of `i` onwards) out of these iterators because I think putting the common code into a helper would require invoking a helper iterator from these iterators, passing params to it in order to get the compiler to specialize it appropriately for both cases and result in the same tight code loop that we're generating for these.  For me, I'm not sure that the bit of code sharing that would result in is worth the finger crossing that we would get the same performance (today and in the future), particularly since the compiler would effectively create the clones again for us.  Put another way, I'd rather have these loops expressed clearly and to see the differences myself by comparing them than to abstract the loop itself away into a shared iterator like that.

But hoisting the bounds checking logic out into a helper seems like a no-brainer.

Let me know if that approach is OK with you or if you see a way of refactoring the loop itself that seems more reasonable than my intuition.

Reply:
> Let me know if that approach is OK with you or if you see a way of refactoring the loop itself that seems more reasonable than my intuition.

I'm fine with that, I don't see duplicate code necessarily as a problem so long as there's some justification for it (e.g. concerns over performance or clarity).

It's a matter of taste too, so do as you best see fit. 

If it were me: I'd probably factor out the common range checks, strip the comments out of the `== BoundedRangeType.boundedHigh` case, then add a comment to that case saying the implementation mirrors the other one and look there for details.

Reply:
https://github.com/chapel-lang/chapel/pull/19991/commits/ab4e39bd8856efb4361297b21ab68822c32aa839 implemented a form of this.